webserver: add support for python scripts

Montellese opened the pull request:
This is based on some work I've started over 2 years ago but when I was almost finished @jimfcarroll had re-written the whole python integration and my changes didn't work anymore (as they were in native Python C). Now I finally got back to it and moved it to the code generator implementation and even added additional functionality.

What this does is basically allow webinterface addons to provide and run python scripts which handle incoming HTTP requests with the help of the newly introduced `xbmchttp` python module.

I'll try to break down the changes and additions I've made:
- This is based on #6201 which is where the first 10 commits come from
- The webserver has barely been touched expect for being able to access the full request URI (including path information and query arguments)
- the IHTTPRequestHandler interface has gotten a new helper method called `GetHostnameAndPort()` which queries and parses the `Host` HTTP header to determine the hostname and port used by the client to call our webserver
- `ILanguageInvoker`, `CLanguageInvokerThread` and `CScriptInvocationManager` have been changed so that `ILanguageInvoker` is stored as a shared pointer. Right now the `ILanguageInvoker` used to run a script is deleted right after the script has finished but this doesn't work for running python scripts from the webserver. Either way using a shared pointer is IMO better anyway since these pointers are passed around between classes in quite a few places.
- I have renamed `CScriptInvocationManager::Execute()` to `ExecuteAsync()` (which returns directly) and added a `ExecuteSync()` which waits for the script to finish or abort it after a specified timeout
- I have added a new macro `SWIG_IMMUTABLE()` to support read-only members/properties in the codegenerator logic
- I have refactored the parsing and handling of python exceptions to be able to pass them on to `CPythonInvoker::onError()`
- I have cleaned up all the python input and output typemap definitions which results in cleaner code generation
- I have added output typemaps for `XBMCAddon::Dictionary<>` and `std::map<>`
- I have added a new type `XBMCAddon::MultiDictionary<>` as a wrapper around `std::multimap<>` with a python output typemap
- I have changed the python output typemaps of `long long` to use `PyLong_FromLongLong()` instead of `Py_BuildValue()`
- I have added a python output typemap for `unsigned long long` (the input typemap already exists)
- I have added support for static class methods to the python code generator
- I have fixed the python code generator for the index operator (`operator[]`)
- I have added support for passing data to the index operator (`operator[]`) to the python code generator
- I have fixed the python code generator for class members of a custom API type (not a builtin one)
- I have added support for python's `tp_str` (`__str__`) which can be used by providing a `toString()` class method
- I have added limited support for python`s`tp_as_sequence`which can be used by providing a`size()`and`contains()`class method. This allows using`if 'foo' in bar` in python scripts.
- I have added limited python input and output typemas for `CVariant`. There's no support for arrays or objects since those resulted in infinite recursive calls to the typemaps. Maybe @jimfcarroll has an idea for this.
- I have added a new `xbmchttp` module which contains constants for HTTP headers and status codes and the following classes: `HttpRequest`, `Cookie` and `Session` (and `Table`). These methods use mostly the same method and property names as Apache's mod_python (see http://modpython.org/). It is not identical but most of the important methods and properties are there.
- The `HttpRequest` class represents the HTTP request and stores all the information to put together the HTTP response.
- The `Cookie` class is a helper class for retrieving, parsing and creating HTTP cookies.
- The `Session` class uses the `CHTTPSessionManager` to remember session objects across HTTP requests and provides the session ID in a cookie named `kodisid` in a HTTP response and supports retrieving the session ID from a cookie named `kodisid` or from a HTTP GET parameter named `sid` from HTTP requests. A session can be created, loaded, invalidated and deleted and can be used as a map to store information.
- I have added `CHTTPPythonInterpreter` (should probably be renamed to `CHTTPPythonInvoker`) as a specialization of `CPythonInvoker`. The main differences are that python's `print` is redirected to the content of the HTTP response, that only the `xbmc` and `xbmchttp` modules are loaded (no sense in providing access to the others), that a `HttpRequest` object with the name `req` is injected into the script's global variable space and that exceptions are turned into a HTML error page.
- Finally I've added a `CHTTPPythonHandler` as a specialization of `IHTTPRequestHandler` which is responsible for setting up the python invoker, executing the script, turning the result into a HTTP response and terminating scripts that run longer than 30 seconds.
- Last but not least I've temporarily added a test python script to the default webinterface to demonstrate some of the available features and how it basically works

I know this is a lot. I'd be grateful if @jimfcarroll (and anyone else who dares) could take a close look at all the changes I made to the python code generator (and the typemaps) and also to the new module and classes I've added. The webserver changes are minimal and the new HTTP request handler is pretty straight forward as well.

Montellese committed:
use shared pointer of ILanguageInvoker

Montellese committed:
CScriptInvocationManager: rename Execute() to ExecuteAsync() and add documentation

Montellese committed:
webserver: add CHTTPPythonInvoker and CHTTPModPythonInvoker to run python scripts from webinterfaces using mod_python

Montellese committed:
webserver: add CHTTPPythonHandler to handle python scripts for webinterfaces using mod_python

Montellese committed:
[win32] update VS project files

Montellese committed:
TEMP: [webinterface.default] add test.py

Montellese committed:
swig: add SWIG_IMMUTABLE macro for read-only properties

Montellese committed:
python: refactor parsing of python exceptions

Montellese committed:
python: add out typemaps for Dictionary<> and std::map<>

Montellese committed:
python: add support for static class methods

Montellese committed:
python: add support for tp_str stringification method

Montellese committed:
python: add MultiDictionary with out typemap

Montellese committed:
CScriptInvocationManager: add ExecuteSync()

Montellese committed:
python: fix code generation for index operator

Montellese committed:
python: add support to the codegenerator for passing data to the index operator using mp_ass_subscript

Montellese committed:
python: fix property getters/setters for custom API types

Montellese committed:
python: add support for tp_as_sequence

Montellese committed:
interfaces: add Session to xbmcmod_python and CHTTPSessionManager

Montellese committed:
webserver: remember the full path URI of a request

Montellese committed:
IHTTPRequestHandler: add GetHostnameAndPort internal helper method

Montellese committed:
webserver: remember the full request URI (not just the path)

Montellese committed:
python: use PyLong_FromLongLong instead of Py_BuildValue() and add output type map for 'unsigned long long'

Montellese committed:
python: cleanup input/output typemap definitions for cleaner code generation

Montellese committed:
python: add limited input/output typemaps for CVariant

topfs2 commented:
Oooh love it. Almost want to say that we could have this instead of the
image hook, as we could do a script using pil.

tamland commented:
Wow that *is* a lot!:) Haven't looked at anything yet but reading your summary have you though about wsgi? If that's possible that would really take this to another level.

Montellese commented:
Never heard of WSGI but I'll look into it.

tamland commented:
It's the standardized way for servers to run python applications. It means all existing frameworks and applications would run out of the box. I don't know the technical details, but after looking at mod_python more closely (which I have never heard of) I think it's similar to that.

jimfcarroll commented:
Wow. Just saw this.

I'll review it tonight.

Montellese commented:
@tamland: I looked a bit at WSGI and it looks very interesting. The problem is that it looks to be completely based on callables (from the server to the application and vice versa) and I'm not sure if our current code generator supports this (and if so to what degree). Furthermore it sounds to (I have no experience so I might be wrong) that everyone uses WSGI through a middleware framework. That would mean that every webinterface would have to provide the middleware framework of their choosing with the webinterface addon.

@jimfcarroll: Thanks, looking forward to it.

tamland commented:
@jimfcarroll Do you know if swig can do that? I think calling *from the server* should be possible at least.

>Furthermore it sounds to (I have no experience so I might be wrong) that everyone uses WSGI through a middleware framework. That would mean that every webinterface would have to provide the middleware framework of their choosing with the webinterface addon.

That is my understanding as well, *if* you want those higher level features, but you can still choose not to. Couldn't you say the same about mod_python and variants though? Except for the latter, no such middleware framework exists yet. (And the frameworks could be provided as addon modules)

Montellese commented:
@tamland: I'm pretty sure that swig can do this but we only use swig for the parsing of the C/C++ implementation. The whole python code generation is done with our own (actually @jimfcarroll's) implementation so it does not support swig's full feature set.

AFAIK right now we only support callbacks from core to python on methods that implement a specific interface.

jimfcarroll commented on xbmc/interfaces/generic/ILanguageInvoker.cpp:
```diff
@@ -0,0 +1,82 @@
+    return;
+
+  m_state = state;
+}
```
Heh. I take it you don't like inlines. :-)

jimfcarroll commented:
@Montellese @tamland I'm not sure what you mean by "does the code generator support 'this'" but I don't see anything in the description that it doesn't do. You can call either way across the scripting language boundary. You can call callbacks implemented in the scripting language or call callbacks from the scripting language (though you may need to write a simple proxy class for the later).

I don't know anything about WSGI but based on a 10 second google, isn't it already a python interface?

tamland commented:
@jimfcarroll In short you need to pass a callable `start_response` that takes two arguments to be called *from* the scripting language. That's what I was referring to. I guess the question was silly as you could just create a special class for that.

Montellese commented:
Since I'll try to look into WSGI more deeply I have rebased this, fixed a bug in the Session logic and moved all the files that belong to the mod_python based implementation into a `mod_python` subdirectory of `xbmc/interfaces/legacy`.

Montellese committed:
addons: add CWebinterface with "type", "library" and "entry" attributes for xbmc.gui.webinterface

Montellese commented:
I've taken another look at WSGI and the specification seems to be pretty basic.

I've solved the callable requirement by adding support for `operator()` and it's Python C equivalent `tp_call` and I can execute a very basic script. But I had to write some stuff in Python C mainly because I need a way for the script to tell me which method to call (the other possibility would be to expect a global function with a specific name like mod_wsgi but it would still require Python C code to get the reference to the function).
Furthermore WSGI requires implementations for an input and an error stream (see https://www.python.org/dev/peps/pep-0333/#input-and-error-streams) which could mostly be written with our code generator except for the `__iter__()` iterator implementation.
Since WSGI is python specific I was wondering if going the Python C approach would be ok? We probably don't ever want to provide this in any other language.

tamland commented:
Very nice! Will try to look at that in details later. I'm not entirely sure what you're referring to as 'the Python C approach', but I see no problem making the wsgi api for python only. In fact I don't think any other language should implement it. E.g java version should implement servlets.

>(the other possibility would be to expect a global function with a specific name like mod_wsgi but it would still require Python C code to get the reference to the function).

After a quick look at other servers, it seems to be configurable on the sever side. It could be an attribute in the addon.xml extension point. So library+callable name.

tamland commented:
Ok, I'm slightly puzzled by how this thing works now.. Where did the 'library' attribute go and how do you even know which module to import (in both mod_python and wsgi version)?

Montellese commented:
What "library" attribute? In the addon.xml of the webinterface you state whether you want to use mod_python or wsgi and depending on that you can either import `xbmcmod_python` or `xbmcwsgi`.

With "Python C approach" I mean that I write the python code myself using `PyObject` and the Python C API. The rest of our API is written in pure C++ and the codegenerator takes care of writing the wrappers/proxies that convert between our "native" C++ classes and types and the Python C API types.

Montellese commented:
Sorry I forgot to mention that the WSGI work is located in a separate branch on top of this one, see https://github.com/Montellese/xbmc/commits/webserver_python_wsgi.

tamland commented:
Yeah, I found the branch and I get the code generator stuff:). Just a slight misunderstanding here about that Python C code you wrote. I imagined it was part of the script execution (called from kodi, not the other way around). I which case, Python C code would be required. (see inline comment)

Another possibly more fundamental problem I just realized is that in mod_python, apache does all the url handling. (Hench the confusion about the library attribute). In WSGI the application does all of that, and there is only one entry point to the application.

tamland commented on xbmc/network/httprequesthandler/HTTPPythonHandler.cpp:
```diff
@@ -0,0 +1,242 @@
+    }
+
+    LanguageInvokerPtr languageInvokerPtr(pythonInvoker);
+    int result = CScriptInvocationManager::Get().ExecuteSync(m_scriptPath, languageInvokerPtr, m_addon, args, 30000, false);
```
Basically, the `xbmcwsgi::handle_request` Python C implemetation goes here, where `PyObject_CallObject` replaces `PyRun_FileExFlags` in `ExecuteSync`

Montellese replied:
The `xbmcwsgi::handle_request` can't directly go here because then you don't have any timeout handling anymore. If anything it should be handled in `CHTTPPythonWsgiInvoker`.
What I'm not sure about is how I can use `PyObject_CallObject` without specifying which python script it should run which is donewith `PyRun_FileExFlags`.

Montellese replied:
Ah I think I just understood. The idea would be to automatically run the WSGI script when the webserver is started and all it does is to register the proper WSGI handler method with the interpreter. Then whenever a request is made to the webinterface the registered WSGI handler method is called with the details of the request. That would also save the time necessary to setup the python interpreter for every HTTP request. Does that sound about right or did I get it wrong?

tamland replied:
Uhm no:) But that you don't have to setup the interpreter for every request (for performance) seem to be one of the things mentioned as motivation in the spec. Though you aren't going to have to handle *that* many requests are you? What I meant was something simple like: https://docs.python.org/2/extending/extending.html#calling-python-functions-from-c

In WSGI one should **not** run the script but instead something like:

1. Do a regular import of the script using `PyImport_Import`
2. Get reference to the callable in that module with `PyObject_GetAttrString`
3. Invoke/get the response from that callable with `PyObject_CallObject`

That's the low level stuff. I have no idea where to put it. It replaces parts of `CPythonInvoker::Execute` but doesn't fit very well with the whole executing a script by file name concept.

Montellese replied:
AFAICT the part about calling python functions from c you linked to assumes an already running script. 
But I think my confusion was that I didn't (and still don't) realize how python handles modules. That `PyImport_Import("test")` automatically load a file called `test.py`in the path? That's what I don't understand right now i.e. how does python know where to look for the request handler callable?

tamland replied:
[This page](https://docs.python.org/3/extending/embedding.html#pure-embedding) is better and contains a complete example. I believe the PyRun_* functions is just a higher level interface for the same thing. Eventually it imports the module in the same way. The only difference between "running" the script and importing it is that the global `__name__` variable is set to `__main__` before import. (But I'm no expert so we may have to call in jimfcarroll on this.) Hope that page clears it up?

topfs2 replied:
I think that's right. Because iirc with Apache you can use different mod
pythons on different paths too, which essentially is the same as for us.
I.e. we hook up wsgi on a certain add-on id path and then use our image
stuff on another.

tamland commented on xbmc/network/httprequesthandler/HTTPPythonHandler.cpp:
```diff
@@ -0,0 +1,242 @@
+    m_responseRanges()
+{
+  // get the real path of the script and check if it actually exists
+  m_response.status = CHTTPWebinterfaceHandler::ResolveUrl(m_request.pathUrl, m_scriptPath, m_addon);
```
And here. In WSGI this path should dispatch to the same application. I.e `m_scriptPath` should always be `m_addon.LibPath()` for instance.

Montellese replied:
I have re-written the URL -> file resolving and extended the webinterface's `addon.xml` to support `library` and `entry` (where `entry` can be used to specify the python callable for WSGI or to specify the default file for static or mod_python webinterfaces).
For WSGI we only support requests to the root path of the server for now. We'll have to look into whether this is enough or if we need a way to also forward other URLs to the WSGI script.

Montellese commented:
We can't pass all URLs to the WSGI script because we also need to handle URLs which are completely independent of the webinterface like VFS, images and JSON-RPC. Furthermore it must still be possible to load images provided by the webinterface and I'm not sure if it makes sense to go through the WSGI script for that because it doesn't have access to our VFS etc.

All in all I might not be the right person to implement this because I have zero experience with WSGI and only very little experience with Python. So I just read the spec and implemented the obvious API requirements.

tamland commented:
I'm learning as I go here too. (I wrote a Flask application once. That's my experience). If there's anything I can help implement I'm happy to do so. I could try to finish the wsgi stuff at least. I think you already completed most of it.

tamland commented:
About static files: only the /addons/addon.id/ path should be handled by the WSGI script. You can still serve everything outside of that as static files right? [Frameworks seems to support serving static files via the application](http://flask.pocoo.org/docs/0.10/quickstart/#static-files). Also the "/static" sub-path of addons could be reserved for static files by the web server for instance as that documentation suggests.

tamland commented on xbmc/network/httprequesthandler/python/Makefile.in:
```diff
@@ -0,0 +1,8 @@
+        HTTPPythonInvoker.cpp \
+        HTTPSessionManager.cpp \
+
+LIB = httprequesthandler-python.a
```
missing s in httprequesthandler**s**-python.a

tamland commented on xbmc/network/httprequesthandler/python/Makefile.in:
```diff
@@ -0,0 +1,8 @@
+        HTTPSessionManager.cpp \
+
+LIB = httprequesthandler-python.a
+
```
needs `INCLUDES += -I/usr/include/python2.7`

tamland commented on xbmc/interfaces/legacy/mod_python/Table.h:
```diff
@@ -0,0 +1,102 @@
+      * key          : string or unicode - the key of an item.\n
+      * default      : [opt] string or unicode - default value returned if the given key doesn't exist.\n
+      */
+      inline String get(const String& key, const String& default = "")
```
default is a keyword... (??)

tamland commented on xbmc/network/httprequesthandler/python/HTTPSessionManager.cpp:
```diff
@@ -0,0 +1,261 @@
+  CSingleLock lock(m_critSection);
+  do
+  {
+    std::string temp = StringUtils::Format("%d%d%d%s", time(NULL) * (10000 + count), rand(), rand(), inet_ntoa(ipInfo->client_addr->sin_addr));
```
‘struct sockaddr’ has no member named ‘sin_addr’. A quick grep showed a lot of casting to `struct sockaddr_in` (I have no idea what any of this is)

tamland commented on xbmc/interfaces/legacy/mod_python/Cookie.cpp:
```diff
@@ -0,0 +1,278 @@
+      request->m_request->responseHeaders.insert(std::make_pair(MHD_HTTP_HEADER_SET_COOKIE, cookie->toString()));
+    }
+
+    XBMCAddon::Dictionary<Cookie*> Cookie::get_cookies(HttpRequest *request)
```
missing throw on these

Montellese committed:
interfaces: add xbmcmod_python module with HttpRequest and Cookie

Montellese commented:
@tamland: Thanks for the review. I hope I have addressed them all. Haven't tried building it for linux yet so that's why I didn't spot those buildsystem issues.

Montellese commented:
Concerning static files: the default webinterface should always be accessible from the root path of the webserver because that's what people type into their browser. What we could do is forward any requests without a path to the WSGI script (if the addon supports it) and everything else is handled by the webserver as it is now.

Montellese committed:
webserver: fix CHTTPWebinterfaceHandler::ResolveUrl() to support a custom entry page

tamland commented:
Figured as much, but when I got those syntax errors I wasn't sure it was supposed to compile yet. You sure it compiles for you:)? Still can't compile it yet. Now I'm stuck at some error in the generated code.

Montellese committed:
codegenerator: improve doxygen integration (thanks notspiff)

Montellese commented:
Thanks @tamland and @notspiff for your fixes. I've added all of them expect for the `CVariant` and `long long` problem which either seems to work on win32 or on linux but not both right now.

Montellese committed:
addons: add new xbmc.webinterface addon

Montellese committed:
python: add input/output type maps for 'unsigned int'

jimfcarroll commented on xbmc/interfaces/python/PythonSwig.cpp.template:
```diff
@@ -117,6 +126,8 @@ void doMethod(Node method, MethodType methodType)
     else
       return;
   }
+  else if (method.@name == "contains")
```
Hi @Montellese,

This seems like an odd special case. If my method name is exactly "contains" then I'm going to generate the method call differently?

jimfcarroll replied:
Okay, I get it. If the class has a 'contains' method then you're going to be able to call it using sequence ops from python. Will this break classes that simply have a method called 'contains' or will it just provide an alternative way of calling it from python?

Montellese replied:
Yeah because I ignore `contains` (and `toString`) in the regular list of methods handled by `doMethod()`, see https://github.com/xbmc/xbmc/pull/6264/files#diff-899036c0c39c2c424042c8102edc523bR379. I checked and there's no `contains` method in the current legacy API.
Do you have a better idea how to decide (maybe more explicit) whether an instance of a class can be used as a sequence?

Montellese replied:
@jimfcarroll: I have changed the logic. Now you need to add
```
%feature("sequence") Session "contains"
```
into the SWIG interface definition file to make the class `Session` usable as a sequence using the `contains` method for `tp_contains`. The method used for `tp_contains` is automatically excluded from the actual method list.
That should make it more explicit.

jimfcarroll commented:
@Montellese , sorry about the first comment. Once I understood the impact (of the code you added for 'contains') I realized it's just a python convenience. I did the same thing for "rich comparison" as well as operator[] so I don't think there's any need to add the feature tag. As a matter of fact it's probably worse to require it unless there's a reason NOT to be able to use this convenience from python

jimfcarroll commented:
BTW, I think this looks good. I didn't really look carefully at the webserver side of the world but I dug through the code generator changes. It's good to see someone in there making changes. @Montellese is probably a Groovy convert now? :-)

Montellese commented:
Silenced a warning and updated .gitignore with the new generated Makefiles.

@jimfcarroll: Well it might still be that someone simply wants to provide a `contains()` method and doesn't realize that it is handled differently. The `operator[]` is a bit different because it can't be used as `operator[]` in python and using `[]` there is the expected behaviour anyway. But I can revert the `%feature("sequence")` again if you think that's better.

jimfcarroll commented:
@Montellese, your call.

Montellese commented:
I'm closing this for now as WSGI is preferred over mod_python (or can be run on mod_python).
